gemini/sample-apps/llamaindex-rag/backend/indexing/docai_parser.py

```diff
@@ -0,0 +1,264 @@
+            raise
+
+    def _wait_for_operations(self, operations, timeout_sec, check_in_interval_sec):
+        time_elapsed = 0
+        while any(not operation.done() for operation in operations):
+            time.sleep(check_in_interval_sec)
+            time_elapsed += check_in_interval_sec
+            if time_elapsed > timeout_sec:
+                raise TimeoutError("Timeout exceeded!")
+
+        # Check for errors in completed operations
+        for operation in operations:
+            if operation.exception():
+                raise KeyError(f"Operation failed: {operation.exception()}")
+
+    def _get_results(self, operations):
+        results = []
+        for operation in operations:
+            metadata = operation.metadata
+            if hasattr(metadata, "individual_process_statuses"):
+                for status in metadata.individual_process_statuses:
+                    results.append(
+                        DocAIParsingResults(
+                            source_path=status.input_gcs_source,
+                            parsed_path=status.output_gcs_destination,
+                        )
+                    )
+            else:
+                print(f"Warning: Unexpected metadata structure: {metadata}")
+        return results
+
+    def _parse_from_results(self, results):
+        documents = []
+        storage_client = storage.Client()
+
+        for result in results:
+            print(
+                f"Processing result: source_path={result.source_path}, \
+                    parsed_path={result.parsed_path}"
+            )
+            if not result.parsed_path:
+                print(
+                    f"Warning: Empty parsed_path for source\
+                          {result.source_path}. Skipping."
+                )
+                continue
+
+            try:
+                bucket_name, prefix = result.parsed_path.replace("gs://", "").split(
+                    "/", 1
+                )
+            except ValueError:
+                print(
+                    f"Error: Invalid parsed_path format for {result.source_path}. Skipping."
+                )
+                continue
+
+            bucket = storage_client.bucket(bucket_name)
+            blobs = list(bucket.list_blobs(prefix=prefix))
+            print(f"Found {len(blobs)} blobs in {result.parsed_path}")
+
+            for blob in blobs:
+                if blob.name.endswith(".json"):
+                    print(f"Processing JSON blob: {blob.name}")
+                    try:
+                        content = blob.download_as_text()
+                        doc_data = json.loads(content)
+
+                        if (
+                            "chunkedDocument" in doc_data
+                            and "chunks" in doc_data["chunkedDocument"]
+                        ):
+                            for chunk in doc_data["chunkedDocument"]["chunks"]:
+                                doc = Document(
```

Comment:
You can probably use Document AI Toolbox to get this easier. There's support for Layout Parser.

https://cloud.google.com/document-ai/docs/toolbox#code_samples